src/lib/core/style/focus-classes.ts

```diff
@@ -29,66 +40,143 @@ export class FocusOriginMonitor {
-      this._lastTouchTarget = null;
-      this._setOriginForCurrentEventQueue('keyboard');
-    }, true);
-
-    // On mousedown record the origin only if there is not touch target, since a mousedown can
-    // happen as a result of a touch event.
-    document.addEventListener('mousedown', () => {
-      if (!this._lastTouchTarget) {
-        this._setOriginForCurrentEventQueue('mouse');
-      }
-    }, true);
-
-    // When the touchstart event fires the focus event is not yet in the event queue. This means we
-    // can't rely on the trick used above (setting timeout of 0ms). Instead we wait 650ms to see if
-    // a focus happens.
-    document.addEventListener('touchstart', (event: Event) => {
-      if (this._touchTimeout != null) {
-        clearTimeout(this._touchTimeout);
-      }
-      this._lastTouchTarget = event.target;
-      this._touchTimeout = setTimeout(() => this._lastTouchTarget = null, TOUCH_BUFFER_MS);
-    }, true);
-
-    // Make a note of when the window regains focus, so we can restore the origin info for the
-    // focused element.
-    window.addEventListener('focus', () => {
-      this._windowFocused = true;
-      setTimeout(() => this._windowFocused = false, 0);
+  /** Weak map of elements being monitored to their info. */
+  private _elementInfo = new WeakMap<Element, MonitoredElementInfo>();
+
+  constructor(private _ngZone: NgZone) {
+    this._ngZone.runOutsideAngular(() => {
+      // Note: we listen to events in the capture phase so we can detect them even if the user stops
+      // propagation.
+
+      // On keydown record the origin and clear any touch event that may be in progress.
+      document.addEventListener('keydown', () => {
+        this._lastTouchTarget = null;
+        this._setOriginForCurrentEventQueue('keyboard');
+      }, true);
+
+      // On mousedown record the origin only if there is not touch target, since a mousedown can
+      // happen as a result of a touch event.
+      document.addEventListener('mousedown', () => {
+        if (!this._lastTouchTarget) {
+          this._setOriginForCurrentEventQueue('mouse');
+        }
+      }, true);
+
+      // When the touchstart event fires the focus event is not yet in the event queue. This means
+      // we can't rely on the trick used above (setting timeout of 0ms). Instead we wait 650ms to
+      // see if a focus happens.
+      document.addEventListener('touchstart', (event: Event) => {
+        if (this._touchTimeout != null) {
+          clearTimeout(this._touchTimeout);
+        }
+        this._lastTouchTarget = event.target;
+        this._touchTimeout = setTimeout(() => this._lastTouchTarget = null, TOUCH_BUFFER_MS);
+      }, true);
+
+      // Make a note of when the window regains focus, so we can restore the origin info for the
+      // focused element.
+      window.addEventListener('focus', () => {
+        this._windowFocused = true;
+        setTimeout(() => this._windowFocused = false, 0);
```

Comment:
Just an optional nit. How about extracting this into its own function?

```ts
constructor(ngZone: NgZone) {
  ngZone.runOutsideAngular(() => this._registerDocumentEvents());
}
```

This way we would get rid of unnecessary indentation and also have it pretty clean.